content/blog/elevate-your-v-project-with-unit-tests.md

```diff
@@ -0,0 +1,749 @@
+[documentation](https://github.com/vlang/v/blob/master/doc/docs.md#testing).
+
+## A Brief Recap of `geo`
+
+In the [last tutorial](/the-complete-beginners-guide-to-cli-apps-in-v/)
+we created a complete command-line application using the built-in `flag`
+module. So we already have the directory structure like this:
+
+```text
+$ tree .
+.
+├── geometry
+│   ├── metadata.v
+│   ├── options.v
+│   ├── quadrilateral.v
+│   └── triangle.v
+├── geo.v
+└── v.mod
+```
+
+The `v.mod` file anchors the root of the project so that it becomes easier
+to use the `geometry` module. It also contains information about the project
+itself, like the version, description, license, and so on.
+
+The `geo.v` contains the main entry point of the application. It processes
+the command-line arguments and calls the `generate_shape` function with the
+appropriate arguments.
+
+The `geometry` directory contains the source for out module spread across
+multiple files. The `metadata.v` file contains the metadata about the module
+that it reads from the `v.mod`. Next, `options.v` defines the various enums
+and helper structs that help standardise the input. Finally, the `triangle.v`
+and `quadrilateral.v` files contain the actual code for generating the
+desired shapes.
+
+We took an architectural decision to keep the signature of all of the
+shape generation functions same:
+`pub fn generate_xyz(options ShapeOptions) []string`. We will make use
+of this to keep our testing strategy simple.
+
+## First Steps
+
+First, we create a `test` directory. Then we create a `unit_tests` directory.
+Inside that, we create a file `triangle_test.v` in which we will write our
+first test function.
+
+The general way to proceed is to create a new `ShapeOptions` struct with the
+necessary options. Then we obtain the result of calling the required
+`generate_xyz()` function. We will then proceed to check if it is the same
+output as we expected. Let's start with the `generate_left_triangle` function.
+
+The final code will be as follows:
+
+```v {linenos=table}
+import geometry
+
+fn test_generate_left_triangle_big_6() {
+	opts := geometry.ShapeOptions{
+		size: 6
+	}
+	shape := geometry.generate_left_triangle(opts)
+	assert shape == ['*', '**', '***', '****', '*****', '******']
+}
+```
+
+Notice that we do not have a `module xyz` file at the top. We do not
+need this line in our case because we're testing public functions.
+However, you may need to test the private (module-only) functions in some
+instances. In that case, the test file is put in the same folder as the
+source file containing the code you want to test, and the test file
+needs to have the module declaration at the top.
+
+You may include a module declaration like `module geometry_test` at
+the top of these files but that is not compulsory, as stated before.
+Ideally, you should only be testing the public facing functions so that
+you have room to refactor your code without any users noticing.
```

Comment:
```suggestion
Ideally, you should be testing the public facing functions so that
you have room to refactor your code without any users noticing.
```

Reply:
```suggestion
Ideally, you should test all public facing functions so that
you have room to refactor your code without any users noticing.
```